src/Features/Core/Portable/DuplicateMethod/AbstractDuplicateMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis.DuplicateMethod
+{
+    internal abstract class AbstractDuplicateMethodCodeRefactoringProvider<TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var method = root.FindNode(context.Span).GetAncestorOrThis<TMethodDeclarationSyntax>();
+            if (method == null)
+            {
+                return;
+            }
+
+            if (method.ContainsDiagnostics)
+            {
+                return;
+            }
+
+            var identifier = GetIdentifier(method);
+            if (context.Span.Length > 0 &&
```

Comment:
❗️ I'm not a fan of offering this refactoring *everywhere* within the method body. At minimum we should require the caret be placed on the line containing `identifier`.